src/sagemaker_xgboost_container/algorithm_mode/train.py

```diff
@@ -28,10 +31,51 @@
 from sagemaker_xgboost_container.callback import add_debugging
 from sagemaker_xgboost_container.constants.xgb_constants import CUSTOMER_ERRORS
 
+FORCE_CHECKPOINT_DIR = "/opt/ml/tmp"
 
 logger = logging.getLogger(__name__)
 
 
+def add_sigterm_handler(model_dir, force_checkpoint_dir, is_master):
+    """Stop training and save model when SIGTERM is received.
+
+    Trained model is only saved if 'force_checkpointing_dir' is not None and if 'is_master' is True,
+    otherwise function terminates without saving model.
+
+    :param model_dir: Directory where model will be saved
+    :param force_checkpoint_dir: Directory where intermediate models are currently saved
+    :param is_master: True if single node training, or the current node is the master node in distributed training
+    """
+    def _terminate():
+        sys.exit(0)
+
+    def _save_file_to_disk(signo, frame):
+        if force_checkpoint_dir and is_master:
+            model_files = [data_file for data_file in os.listdir(model_dir)
+                           if os.path.isfile(os.path.join(model_dir, data_file))]
+
+            if len(model_files) == 0:
+                checkpoint_files = [data_file for data_file in os.listdir(force_checkpoint_dir)
+                                    if os.path.isfile(os.path.join(force_checkpoint_dir, data_file))]
+                file_name = train_utils.get_latest_checkpoint(checkpoint_files)
+
+                if file_name:
+                    current_file_path = os.path.join(force_checkpoint_dir, file_name)
+                    new_file_path = os.path.join(model_dir, file_name)
+
+                    try:
+                        shutil.move(current_file_path, new_file_path)
+                        logging.debug("Stored intermediate trained model at {}".format(new_file_path))
+                    except shutil.Error:
+                        logging.debug("Failed to move intermediate model from {} to {}"
+                                      .format(current_file_path, new_file_path))
+                        pass
+
+        _terminate()
```

Comment:
I see a bunch of nested ifs here. Is it possible to simplify this? Perhaps make it more modular by moving some of the logic to its own function?

Reply:
Changed in new commit by creating function in train_utils.